server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -677,19 +795,44 @@ private static boolean removedNodesCleanupNeeded(SnapshotsInProgress snapshotsIn
     }
 
     /**
-     * Finalizes the shard in repository and then removes it from cluster state
-     * <p>
-     * This is non-blocking method that runs on a thread from SNAPSHOT thread pool
+     * Finalizes the snapshot in the repository.
      *
      * @param entry snapshot
      */
-    private void endSnapshot(SnapshotsInProgress.Entry entry, Metadata metadata) {
-        if (endingSnapshots.add(entry.snapshot()) == false) {
-            return;
+    private void endSnapshot(SnapshotsInProgress.Entry entry, Metadata metadata, @Nullable RepositoryData repositoryData) {
+        final boolean newFinalization = endingSnapshots.add(entry.snapshot());
+        final String repoName = entry.repository();
+        synchronized (currentlyFinalizing) {
```

Comment:
This synchronization on `currentlyFinalizing` in a bunch of spots is not great admittedly but for step 1 I traded efficiency for safety in many spots like this one.
I have some planned simplifications that will move all the logic for triggering deletes and finalizations exclusively to the master update thread (most of them already are) which will remove the need for most synchronization on the fields in `SnapshotsService`.